s3transfer/upload.py

```diff
@@ -286,6 +286,157 @@ def _get_put_object_fileobj(self, fileobj):
         return fileobj
 
 
+class UploadNonSeekableInputManager(UploadInputManager):
+    """Upload utility for a file-like object that cannot seek."""
+    def __init__(self, osutil, transfer_coordinator):
+        super(UploadNonSeekableInputManager, self).__init__(
+            osutil, transfer_coordinator)
+        self._initial_data = None
+
+    @classmethod
+    def is_compatible(cls, upload_source):
+        return not seekable(upload_source)
+
+    def stores_body_in_memory(self, operation_name):
+        return True
+
+    def provide_transfer_size(self, transfer_future):
+        # No-op because there is no way to do this short of reading the entire
+        # body into memory.
+        return
+
+    def requires_multipart_upload(self, transfer_future, config):
+        # This is tricky to determine in this case because we can't know how
+        # large the input is. So to figure it out, we read data into memory
+        # up until the threshold and compare how much data was actually read
+        # against the threshold.
+        fileobj = transfer_future.meta.call_args.fileobj
```

Comment:
If the user provided a size themselves, I would say we do not stream the file in. So I would make sure you have this check at beginning: https://github.com/boto/s3transfer/blob/develop/s3transfer/upload.py#L186 and then do the streaming.